src/pages/dashboard/ArrivalByTimeChart/ArrivalByTimeChart.tsx

```diff
@@ -34,128 +45,102 @@ export default function ArrivalByTimeChart({
     current: number
     max: number
     percent: number | null
-    gtfs_route_date: string
-    gtfs_route_hour: string
+    // iso string, can be sorted. TODO: better type definitions
+    gtfs_route_date?: string
+    gtfs_route_hour?: string
   }[]
   operatorId: string
 }) {
-  data = useMemo(() => {
-    let newData = data.filter((item) => !operatorId || item.id === operatorId)
-    if (newData[0]?.gtfs_route_hour) {
-      const fillerArray = []
-      for (const currentId of new Set(newData.map((i) => parseInt(i.id)))) {
-        const allPoints = newData
-          .filter((e) => parseInt(e.id) === currentId)
-          .sort((a, b) => (moment(a.gtfs_route_hour).isAfter(b.gtfs_route_hour) ? 1 : -1))
-        if (allPoints[0]) {
-          // Start gap
-          for (
-            let startHour = moment(allPoints[0].gtfs_route_hour).hour();
-            startHour > 0;
-            startHour--
-          ) {
-            fillerArray.push({
-              ...allPoints[0],
-              max: 0,
-              current: 0,
-              percent: null,
-              gtfs_route_hour: moment(allPoints[0].gtfs_route_hour)
-                .subtract(startHour, 'hours')
-                .toISOString(),
-            })
-          }
-          // During gap
-          let lastHour = allPoints[0].gtfs_route_hour
-          allPoints.forEach((entry) => {
-            for (
-              let delta = moment(entry.gtfs_route_hour).diff(moment(lastHour), 'hours');
-              delta > 1;
-              delta--
-            ) {
-              const currentHour = moment(lastHour).add(1, 'hours').local().toISOString()
-              fillerArray.push({
-                ...entry,
-                percent: null,
-                current: 0,
-                max: 0,
-                gtfs_route_hour: currentHour,
-              })
-              lastHour = currentHour
-            }
-            lastHour = entry.gtfs_route_hour
-          })
+  if (operatorId) data = data.filter((item) => item.id === operatorId)
+  const dataByOperator = useMemo(() => {
+    // find the min and max time, to create a range of dates / hours for the x axis
+    const times = data.map((item) => item.gtfs_route_date ?? item.gtfs_route_hour!).filter(Boolean)
+    if (times.length === 0) return []
+    const minTime = times.reduce((a, b) => (a < b ? a : b))
+    const maxTime = times.reduce((a, b) => (a > b ? a : b))
+
+    const range = getRange(minTime, maxTime, data[0].gtfs_route_hour ? 'hour' : 'day')
+
+    const pointsPerOperator = arrayGroup(data, (item) => item.id).map((operatorData) => {
```

Comment:
I've been braking my head over this `arrayGroup` function, can you help me understand what it does?

Reply:
Hi, sure! I think I tried implementing something similar to this one:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy
But I see it returns an array, instead of an object.
if my data seems like 
```
[{id: 1, fruit: "banana"}, {id: 2, fruit: "apple"}, {id: 2, fruit: "orange"}]
```
I expect `arrayGroup(data, (item) => item.id)` to give me:
```
[
     [{id: 1, fruit: "banana"}],
    [{id: 2, fruit: "apple"}, {id: 2, fruit: "orange"}],
]
```
Let me know if that's ok